concordium-consensus/src/Concordium/GlobalState/Persistent/BlobStore.hs

```diff
@@ -215,6 +215,24 @@ runBlobStoreTemp dir a = bracket openf closef usef
             freeCallbacks bscLoadCallback bscStoreCallback
             return res
 
+-- | Truncate the blob store after the blob stored at the given offset. The blob should not be
+-- corrupted (i.e., its size header should be readable, and its size should match the size header).
+truncateBlobStore :: BlobStoreAccess -> BlobRef a -> IO ()
+truncateBlobStore BlobStoreAccess{..} (BlobRef offset) = mask $ \restore -> do
+  bh@BlobHandle{..} <- takeMVar blobStoreFile
+  eres <- try $ restore $ do
+    hSeek bhHandle AbsoluteSeek (fromIntegral offset)
+    esize <- decode <$> BS.hGet bhHandle 8
+    case esize :: Either String Word64 of
+      Right size -> do
+        let newSize = offset + 8 + size
+        hSetFileSize bhHandle $ fromIntegral newSize
+        putMVar blobStoreFile bh{bhSize = fromIntegral newSize, bhAtEnd=False}
+      _ -> throwIO $ userError "Cannot truncate the blob store: cannot obtain the last blob size"
+  case eres :: Either SomeException () of
+    Left e -> throwIO e
+    Right () -> return ()
```

Comment:
The typical pattern that we use `mask` for is to ensure that we restore the `MVar` in the event of an exception, including an async exception. That is not happening in this case. I think it's probably reasonable to remove the `mask`ing in this case, because if there is some issue from `hSetFileSize` then we can't really expect to recover. (Alternatively, it might make sense to but an error value into the `MVar`, but I'm not sure that would be better.)

As a second point, I think the `blobStoreMMap` should be remapped on truncation. Otherwise, there could potentially be an issue of incorrect data being read.